### PR TITLE
feat: add visualization for no detection area

### PR DESCRIPTION
**Pull request by TomohitoAndo**
Signed-off-by: Tomohito Ando <tomohito.ando@tier4.jp>

## Description

Added visualization for the polygon tagged as `no_detection_area`

Related PR:
- https://github.com/autowarefoundation/autoware.universe/pull/1530

image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: tmp/lanelet2_extension/lib/query.cpp [hunk 1]
```diff
@@ -879,6 +879,19 @@
   return lanelet_sequences_vec;
 }
 
+lanelet::ConstPolygons3d query::getAllNoDetectionArea(
+  const lanelet::LaneletMapConstPtr & lanelet_map_ptr)
+{
+  lanelet::ConstPolygons3d no_detection_area;
+  for (const auto & poly : lanelet_map_ptr->polygonLayer) {
+    const std::string type = poly.attributeOr(lanelet::AttributeName::Type, "none");
+    if (type.compare("no_detection_area") == 0) {
```
[the remark this answers is not in the document]
Comment (TomohitoAndo) [reply tmp/lanelet2_extension/lib/query.cpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/92/commits/4b42e2a3e5396779be10f4cbdee753f533133a31.

[the remark this answers is not in the document]
Comment (TomohitoAndo) [reply tmp/lanelet2_extension/lib/query.cpp]: fixed in https://github.com/autowarefoundation/autoware_common/pull/92/commits/4b42e2a3e5396779be10f4cbdee753f533133a31.

File: tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp [hunk 2]
```diff
@@ -175,6 +175,9 @@ lanelet::ConstLineStrings3d getLinkedParkingSpaces(
 lanelet::ConstLanelets getLinkedLanelets(
   const lanelet::ConstPolygon3d & parking_lot, const lanelet::ConstLanelets & all_road_lanelets);
 
+// query all no detection area in lanelet2 map
+lanelet::ConstPolygons3d getAllNoDetectionArea(const lanelet::LaneletMapConstPtr & lanelet_map_ptr);
+
```
Comment (shmpwk) [inline tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp:180]: [minor comment]

Maybe this function is better placed at L117-L122

https://github.com/autowarefoundation/autoware_common/blob/4b42e2a3e5396779be10f4cbdee753f533133a31/tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp#L117-L122

File: tmp/lanelet2_extension/lib/query.cpp [hunk 3]
```diff
@@ -879,6 +879,19 @@ std::vector<lanelet::ConstLanelets> query::getPrecedingLaneletSequences(
...
+{
+  lanelet::ConstPolygons3d no_detection_area;
+  for (const auto & poly : lanelet_map_ptr->polygonLayer) {
+    const std::string type = poly.attributeOr(lanelet::AttributeName::Type, "none");
+    if (type == "no_detection_area") {
+      no_detection_area.push_back(poly);
+    }
+  }
+  return no_detection_area;
+}
```
Comment (shmpwk) [inline tmp/lanelet2_extension/lib/query.cpp:893]: [minor comment]
Better to put this in the same position as hpp.

File: tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp (hunk 2 above)
Comment (TomohitoAndo) [reply tmp/lanelet2_extension/include/lanelet2_extension/utility/query.hpp]: Thank you for your comments!
I fixed the position of the function.
https://github.com/autowarefoundation/autoware_common/pull/92/commits/99feb65928eb3d9bda1629909d023e005b9620f7

File: tmp/lanelet2_extension/lib/query.cpp (hunk 3 above)
Comment (TomohitoAndo) [reply tmp/lanelet2_extension/lib/query.cpp]: fixed the position of the function.
https://github.com/autowarefoundation/autoware_common/pull/92/commits/99feb65928eb3d9bda1629909d023e005b9620f7

**Review by 0x126 — DISMISSED** [review]
LGTM

**Review by shmpwk — DISMISSED** [review]
LGTM

**Review by kenji-miyake — APPROVED** [review]

**Comment by TomohitoAndo** [comment]
I discussed with @yukkysaito and some changes are needed.
- [ ] change the name of `no_detection_area` to `no_obstacle_segmentation_area` and `no_obstacle_segmentation_area_for_run_out`
- [ ] add library to be able to get lanelet polygon with any tag
- [ ] add documentation for `no_obstacle_segmentation_area` and `no_obstacle_segmentation_area_for_run_out`

**Comment by TomohitoAndo** [comment]
@yukkysaito 
I added changes in this comment. could you please review?
https://github.com/autowarefoundation/autoware_common/pull/92#issuecomment-1220532184

**Review by yukkysaito — APPROVED** [review]
LGTM

**Comment by TomohitoAndo** [comment]
@yukkysaito 
Could you approve runnning the workflow?
image

**Comment by TomohitoAndo** [comment]
@yukkysaito 
Thanks!
Could you merge this PR because I don't have the write access?
image

**Pull request merged**